h2o-algos/src/main/java/hex/rulefit/RuleFit.java

```diff
@@ -307,14 +307,53 @@ SharedTreeModel trainTreeModel(RuleFitModel.Algorithm algorithm, int maxDepth) {
            if (deviance.length < 5) {
                bestLambdaIndex = deviance.length - 1;
            } else {
-               bestLambdaIndex = getBestLambdaIndex(deviance, lambdas);
+               bestLambdaIndex = getBestLambdaIndex(deviance);
+               if (bestLambdaIndex >= lambdas.length) {
+                   bestLambdaIndex = getBestLambdaIndexCornerCase(deviance, lambdas);
+               }
            }
 
             lambdaModel.remove();
             return new double[]{lambdas[bestLambdaIndex]};
         }
 
-        int getBestLambdaIndex(double[] deviance, double[] lambdas) {
+        int getBestLambdaIndex(double[] deviance) {
+            int bestLambdaIndex = deviance.length - 1;
+            if (deviance.length >= 5) {
+                double[] array = differenceArray(signArray(differenceArray(differenceArray(deviance))));
+                for (int i = 0; i < array.length; i++) {
+                    if (array[i] != 0 && i > 0) {
+                        bestLambdaIndex = 3 * i;
+                        break;
+                    }
+                }
+            }
+            return bestLambdaIndex;
+        }
+
+
+        double[] differenceArray(double[] array) {
```

Comment:
Again, if needed, could go to ArrayUtils.

Reply:
moved 3522918